### PR TITLE
feat: validate blocks with replay set, clear replay set

### PR DESCRIPTION
**Pull request by hstove**
Opening this PR to get eyes on this implementation. This PR builds on top of both https://github.com/stacks-network/stacks-core/pull/6020 and https://github.com/stacks-network/stacks-core/pull/6077. To keep this PR "clean", the base branch for is a combination of those two PRs.

This PR implements two main things:

- Validating that a proposed block contains the replay set (if we're in replay mode)
- Clearing the replay set if we're "done" with transaction replay

Knowing if we should be "done" with replay is tricky. What this implementation does is add a field to `BlockValidateOk` with a hash of the replay set. This field is only present if the block was validated against a replay set _and_ it exhausted the replay set. Basically, it's saying "after this block, we can move out of replay". The signer stores this (block_hash, replay_set_exhausted) tuple. Then, in `stacks_block_arrival`, the signer state machine checks if we should move out of replay based on this lookup.

I've added a test (`tx_replay_e2e_test`) which is mainly @jferrant's code from https://github.com/stacks-network/stacks-core/pull/6031/files#diff-9fbead3ef50a4b19fa3c1ce9c11952010f8506706db040df2d85459fe8d518b7 demonstrating:

- bitcoin forks, we go into replay mode
- miner tries to make a non-replay block, gets rejected
- miner makes a replay block, gets approved
- replay set is cleared after
- regular mining resumes

This PR is definitely missing some edge cases, and likely will break other tests, but I wanted to get some eyes on it first.

### CODE REVIEW

File: stacks-signer/src/v0/signer.rs [hunk 1]
```diff
@@ -1588,7 +1600,20 @@ impl Signer {
                 debug!("{self}: Cannot confirm that we have processed parent, but we've waited proposal_wait_for_parent_time, will submit proposal");
             }
         }
-        match stacks_client.submit_block_for_validation(block.clone()) {
+        let is_block_found = block.txs.iter().all(|tx| {
+            matches!(
+                tx.payload,
+                TransactionPayload::Coinbase(..) | TransactionPayload::TenureChange(..)
```
Comment (jferrant) [inline stacks-signer/src/v0/signer.rs:1609]: I am not sure this is sufficient. Is a miner not allowed to tenure extend and include transactions from the relay block in that extend? I mean this could very well be the case in which case, we will need to make sure to update on the miner side.

I think for time based tenure extends, we allow the miner to also add transactions to it (rather than an early exit) and I expected consistent behaviour to do that even during tx replay but maybe it shouldn't ...so not sure, but we will have to make a choice.

**Comment by jferrant** [comment]
Really the one comment which I don't thikn will be hard to resolve. its looking good to me (though will have to update the reorg test during fork and the other failing tests)

File: stacks-signer/src/v0/signer.rs (hunk 1 above)
Comment (hstove) [reply stacks-signer/src/v0/signer.rs:1609]: >  Is a miner not allowed to tenure extend and include transactions from the relay block in that extend?

Yeah, I'm not actually sure. Additionally, we have this ticket for doing [a TenureExtend when the cost limit is reached during replay](https://github.com/stacks-network/stacks-core/issues/5972), so theoretically we shouldn't even be doing time-based extends during replay. Either way, we haven't implemented that on the miner side yet.

I'm definitely OK with allowing other txs during a tenure extend, in which case we'll have to handle that here.

**Comment by hstove** [comment]
Closing in favor of https://github.com/stacks-network/stacks-core/pull/6109
